### PR TITLE
[SPARK-17492] [SQL] Fix Reading Cataloged Data Sources without Extending SchemaRelationProvider

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?

For data sources without extending `SchemaRelationProvider`, we expect users to not specify schemas when they creating tables. If the schema is input from users, an exception is issued. 

Since Spark 2.1, for any data source, to avoid infer the schema every time, we store the schema in the metastore catalog. Thus, when reading a cataloged data source table, the schema could be read from metastore catalog. In this case, we also got an exception. For example, 

``` Scala
sql(
  s"""
     |CREATE TABLE relationProvierWithSchema
     |USING org.apache.spark.sql.sources.SimpleScanSource
     |OPTIONS (
     |  From '1',
     |  To '10'
     |)
   """.stripMargin)
spark.table(tableName).show()
```

```
org.apache.spark.sql.sources.SimpleScanSource does not allow user-specified schemas.;
```

This PR is to fix the above issue. When building a data source, we introduce a flag `isSchemaFromUsers` to indicate whether the schema is really input from users. If true, we issue an exception. Otherwise, we will call the `createRelation` of `RelationProvider` to generate the `BaseRelation`, in which it contains the actual schema.
### How was this patch tested?

Added a few cases.

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @yhuai @cloud-fan

**Comment by cloud-fan** [comment]
ah good catch! But adding a new flag looks a little tricky, let me think if there is better way to fix it

**Comment by gatorsmile** [comment]
@cloud-fan JDBC is also affected by this bug. Do you have any better idea about this issue? Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 1]
```diff
@@ -318,24 +319,28 @@ case class DataSource(
...
         dataSource.createRelation(sparkSession.sqlContext, caseInsensitiveOptions)
       case (_: SchemaRelationProvider, None) =>
         throw new AnalysisException(s"A schema needs to be specified when using $className.")
-      case (_: RelationProvider, Some(_)) =>
-        throw new AnalysisException(s"$className does not allow user-specified schemas.")
+      case (dataSource: RelationProvider, Some(_)) =>
+        if (isSchemaFromUsers) {
+          throw new AnalysisException(s"$className does not allow user-specified schemas.")
+        } else {
+          dataSource.createRelation(sparkSession.sqlContext, caseInsensitiveOptions)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: If the `RelationProvider` doesn't allow user-specified schema, can we assume it's cheap to infer schema for it? Then we can simply check if the given schema matches the schema of relation returned by `createRelation`

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: Yeah, this is a pretty good idea. Let me try it. Thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 2]
```diff
@@ -65,6 +65,26 @@ class InsertSuite extends DataSourceTest with SharedSQLContext {
     )
   }
 
+  test("insert into a temp view that does not point to an insertable data source") {
+    import testImplicits._
+    withTempView("t1", "t2") {
+      sql(
+        """
+          |CREATE TEMPORARY TABLE t1
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala]: let's use CREATE TEMPORARY VIEW

File: sql/core/src/test/scala/org/apache/spark/sql/sources/TableScanSuite.scala [hunk 3]
```diff
@@ -345,34 +345,72 @@ class TableScanSuite extends DataSourceTest with SharedSQLContext {
       (1 to 10).map(Row(_)).toSeq)
   }
 
+  test("create a temp table that does not have a path in the option") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/TableScanSuite.scala]: `temp view`

File: sql/core/src/test/scala/org/apache/spark/sql/sources/TableScanSuite.scala [hunk 4]
```diff
@@ -345,34 +345,72 @@ class TableScanSuite extends DataSourceTest with SharedSQLContext {
...
+    Seq("TEMPORARY VIEW", "TABLE").foreach { tableType =>
+      val tableName = "relationProvierWithSchema"
+      withTable(tableName) {
+        sql(
+          s"""
+             |CREATE $tableType $tableName
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/TableScanSuite.scala]: what does this test?

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala [hunk 5]
```diff
@@ -293,6 +293,39 @@ class DataFrameReaderWriterSuite extends QueryTest with SharedSQLContext with Be
...
+      .option("from", "1")
+      .option("TO", "10")
+      .format("org.apache.spark.sql.sources.SimpleScanSource")
+
+    // when users do not specify the schema
+    checkAnswer(dfReader.load(), spark.range(1, 11).toDF())
+
+    // when users specify the schema
+    val inputSchema = new StructType().add("s", IntegerType, nullable = false)
+    val e = intercept[AnalysisException] { dfReader.schema(inputSchema).load() }
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: there is not test for this case before?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 6]
```diff
@@ -327,8 +327,13 @@ case class DataSource(
         dataSource.createRelation(sparkSession.sqlContext, caseInsensitiveOptions)
       case (_: SchemaRelationProvider, None) =>
         throw new AnalysisException(s"A schema needs to be specified when using $className.")
-      case (_: RelationProvider, Some(_)) =>
-        throw new AnalysisException(s"$className does not allow user-specified schemas.")
+      case (dataSource: RelationProvider, Some(schema)) =>
+        val baseRelation =
+          dataSource.createRelation(sparkSession.sqlContext, caseInsensitiveOptions)
+        if (baseRelation.schema != schema) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: cc @yhuai @liancheng to confirm, is it safe?

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala (hunk 5 above)
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: For `DataFrameReader` APIs, we do not have such a test case

File: sql/core/src/test/scala/org/apache/spark/sql/sources/TableScanSuite.scala (hunk 4 above)
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/TableScanSuite.scala]: My original thought is to provide a comprehensive test coverage of data source table creation/insertion with `path`. This is not related to this PR. Let me get rid of it.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 7]
```diff
@@ -65,6 +65,26 @@ class InsertSuite extends DataSourceTest with SharedSQLContext {
     )
   }
 
+  test("insert into a temp view that does not point to an insertable data source") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala]: hmm, is this test related to this PR?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/TableScanSuite.scala [hunk 8]
```diff
@@ -348,31 +348,51 @@ class TableScanSuite extends DataSourceTest with SharedSQLContext {
...
-      sql(
-        """
-          |CREATE TEMPORARY VIEW relationProvierWithSchema (i int)
-          |USING org.apache.spark.sql.sources.SimpleScanSource
-          |OPTIONS (
-          |  From '1',
-          |  To '10'
-          |)
-        """.stripMargin)
+    Seq("TEMPORARY VIEW", "TABLE").foreach { tableType =>
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/TableScanSuite.scala]: these changed are also not related to this PR right? We are improving the test coverage here.

Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/TableScanSuite.scala]: Yeah

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala (hunk 7 above)
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala]: Not related. This is also to improve the test case coverage. Feel free to let me know if you want to remove it from this PR. Thanks!

**Comment by cloud-fan** [comment]
is this a problem in 2.0?

**Comment by gatorsmile** [comment]
This is a new issue of Spark 2.1, after we physically store the inferred schema in the metastore. 

BTW, I also ran the test cases in Spark 2.0. It works well.

**Comment by cloud-fan** [comment]
thanks, merging to master!
